add rcutils_unsigned_char_array_t

Karsten1987 opened the pull request:
This data structure will eventually be used for serialized messages and replaces the `rcutils_char_array_t` for binary data buffers.
initial issue here: https://github.com/ros2/demos/issues/279

CI:

[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5477)](https://ci.ros2.org/job/ci_linux/5477/)
[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2156)](https://ci.ros2.org/job/ci_linux-aarch64/2156/)
[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4528)](https://ci.ros2.org/job/ci_osx/4528/)
[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5390)](https://ci.ros2.org/job/ci_windows/5390/)

wjwwood commented:
I'm not up to speed on this issue, but what about using `uint8_t` rather than `unsigned char`? There might be a reason not to do so, but it seems like the least ambiguous type to me for serialized data.

Karsten1987 commented:
I am open to new types and names here. I thought unsigned char is just a typedef to uint8_t and thus chose `rcutils_unsigned_char_array_t` to make the contrast to `rcutils_signed_char_array_t` clear. 

Is there any reason to prefer `uin8_t` over `unsigned char` ?

clalancette commented:
> Is there any reason to prefer `uin8_t` over `unsigned char` ?

There is one (pretty pedantic) reason that @sloretz brought up; `char` is not guaranteed to be 8-bits by the standard.  All implementations that we care about make it 8 bits, of course, but naming it `uint8_t` just removes all possible ambiguity.  So I'd agree with @wjwwood on naming it `uint8_array_t`.

wjwwood commented:
> `char` is not guaranteed to be 8-bits by the standard.

The question then becomes, if `char` is not 8-bit, then should our serialized data be an array of 8-bit values or what ever the size of `char` is? 😁

It's a bit academic, because I don't think it will "just" work on a machine where `char` is not 8-bit, and I think we're not likely to find such a machine any time soon. But I do think `uint8_t` is just less ambiguous.

allenh1 commented:
Is there any reason not to call this `rcutils_byte_array_t`?

wjwwood commented:
`rcutils_byte_array_t` would also be fine as it is a commonly used term (`octet` is another example of a term that isn't a type), but byte isn't a C/C++ type so I'd give slight preference to using an explicit type name.

Not that it would require a change in the name of this struct, but MISRA requires that you use explicitly sized types, like `int32` or `int64` rather than `int`, and under the same idea or guidance, using an explicitly sized type in the name here makes sense to me.

allenh1 commented:
@wjwwood that makes sense -- and +1 for the `uint8_t` idea.

Karsten1987 commented:
renamed to `rcutils_uint8_array`

CI:

[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4533)](https://ci.ros2.org/job/ci_osx/4533/)
[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5489)](https://ci.ros2.org/job/ci_linux/5489/)
[![Build Status](https://ci.ros2.org/job/ci_linux-aarch64/2161/badge/icon)](https://ci.ros2.org/job/ci_linux-aarch64/2161/)
[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5395)](https://ci.ros2.org/job/ci_windows/5395/)

Karsten1987 commented:
@ros2/team friendly ping :)

wjwwood commented on include/rcutils/types/uint8_array.h:
```diff
@@ -0,0 +1,109 @@
+ * Be aware, that this will deallocate the memory and therefore invalidates any
+ * pointers to this storage.
+ * If the new size is larger, new memory is getting allocated and the existing
+ * content is copied over.
```
I see now that you are, so this applies I believe:

> the original pointer ptr is invalidated and any access to it is undefined behavior (even if reallocation was in-place).
-- https://en.cppreference.com/w/cpp/memory/c/realloc

Karsten1987 replied:
yes, the pointer might change. Is there any way around that?

wjwwood replied:
> Is there any way around that?

In which case? In general, no I don't think so.

Karsten1987 replied:
We have it documented just before this sentence that all pointers to the underlying buffer will be invalidated. What do you advise to do here?

wjwwood replied:
It says that will only happen "if the new size is larger", which is not the only case and the reason I brought it up initially.

At least it should be stated that you should always assume the pointer is invalid after calling this, regardless of the input arguments and how they related to current values.

wjwwood commented on include/rcutils/types/uint8_array.h:
```diff
@@ -0,0 +1,109 @@
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+rcutils_ret_t
+rcutils_uint8_array_resize(rcutils_uint8_array_t * uint8_array, size_t new_size);
```
> If new_size is zero, the behavior is implementation defined: null pointer may be returned (in which case the old memory block may or may not be freed) or some non-null pointer may be returned that may not be used to access storage.
-- https://en.cppreference.com/w/cpp/memory/c/realloc

Karsten1987 replied:
Do you prefer to strictly forbid to resize to `0`?

wjwwood replied:
I don't know, that might be safest, or you could have it free the whole thing, but that's what fini is for. It's up to you, but I do think blindly passing `0` to `realloc` is unsafe given the uncertainty about what it will end up doing.

jacobperron commented on src/uint8_array.c:
```diff
@@ -0,0 +1,116 @@
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    uint8_array,
+    "uint8 array pointer is null",
+    return RCUTILS_RET_ERROR);
```
`RCUTILS_CHECK_ARGUMENT_FOR_NULL(uint8_array, RCUTILS_RET_INVALID_ARGUMENT);` instead?

Karsten1987 commented:
thanks for the feedback. Given that this `uint8_array` is a direct copy of the `char_array` I believe the same comments apply to these files as well.
Any opinion whether I should fix them in this PR or have a follow up PR on the char array?

wjwwood commented:
> Any opinion whether I should fix them in this PR or have a follow up PR on the char array?

I think it doesn't make sense to merge this with known issues and fix it later, so I'd go with fix in place, retest and merge, then follow up on char array.

Whether or not you follow up on char array in this pr or in a new one doesn't matter to me.

Karsten1987 commented on include/rcutils/types/uint8_array.h:
```diff
@@ -0,0 +1,109 @@
+ * Cleans up and deallocates any resources used in a rcutils_uint8_array_t.
+ * Passing a rcutils_uint8_array_t which has not been zero initialized using
+ * rcutils_get_zero_initialized_uint8_array() to this function is undefined
+ * behavior.
```
maybe it's just a language thing I don't see, but I believe it makes sense to note this here.
if the passed in struct is not correctly initialized, e.g. with `rcutils_get_zero_initialized_uint8_array()` this call to `fini()` is undefined behavior.

wjwwood replied:
But a struct that has been passed to `rcutils_uint8_array_init()` may no longer be zero-initialized (not all of its fields will be `0`), and so it would be in violation of this sentence.

It would be more correct to say, in my opinion, that "the array passed to this function needs to have been initialized by `rcutils_uint8_array_init()` ...".

In any case this sentence didn't make sense to me and I have an intimate understanding of these libraries, so it probably could be made clearer.

Karsten1987 commented:
Hopefully addressed all comments. I decided to iterate over the changes here at first and then apply the same to `rcutils_char_array` in a new PR once this PR is in.

wjwwood commented on include/rcutils/types/uint8_array.h:
```diff
@@ -0,0 +1,109 @@
+ * \param allocator the allocator to use for the memory allocation
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENTS` if any of the arguments are incorrect, or
+ * \return 'RCUTILS_RET_BAD_ALLOC` if no memory could be allocated correctly
```
You might want to leave the `RCUTILS_RET_ERROR`, just in case we need to report an error without breaking API. I think that's pretty common practice to allow all functions with return codes to return the generic error one.

wjwwood commented on include/rcutils/types/uint8_array.h:
```diff
@@ -0,0 +1,109 @@
+ * \param buffer_capacity the size of the memory to allocate for the byte stream
+ * \param allocator the allocator to use for the memory allocation
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENTS` if any of the arguments are incorrect, or
```
```suggestion
 * \return `RCUTILS_RET_INVALID_ARGUMENTS` if any arguments are invalid, or
```

wjwwood commented on include/rcutils/types/uint8_array.h:
```diff
@@ -0,0 +1,109 @@
+/**
+ * Cleans up and deallocates any resources used in a rcutils_uint8_array_t.
+ * The array passed to this function needs to have been initialized by
+ * rcutils_uint8_array_init() or similar.
```
```suggestion
 * rcutils_uint8_array_init().
```

That's the only function right?

wjwwood commented on include/rcutils/types/uint8_array.h:
```diff
@@ -0,0 +1,109 @@
+ *
+ * \param uint8_array pointer to the rcutils_uint8_array_t to be cleaned up
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENTS` if the uint8_array or any of its members is wrong
```
```suggestion
 * \return `RCUTILS_RET_INVALID_ARGUMENTS` if the uint8_array argument is invalid
```

wjwwood commented on src/uint8_array.c:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  uint8_array->buffer_length = 0;
+  uint8_array->buffer_capacity = buffer_capacity;
```
Was this one addressed?

Karsten1987 replied:
The length and capacity are getting set to zero if allocation failed.
``` c++
    RCUTILS_CHECK_FOR_NULL_WITH_MSG(
      uint8_array->buffer,
      "failed to allocate memory for uint8 array",
      uint8_array->buffer_capacity = 0lu;
      uint8_array->buffer_length = 0lu;
      return RCUTILS_RET_BAD_ALLOC);
```
Same applies for `resize()`.

Karsten1987 commented:
@wjwwood do you mind checking it again? I would like to have this in soon so I can build on top of that for the serialized messages.

Karsten1987 commented:
thanks for the review!
